jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * (C) Copyright 2018-2018, by Semen Chudakov and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.shortestpath;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.alg.util.Triple;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of the parallel version of the delta-stepping algorithm.
+ *
+ * <p>
+ * The algorithm is described in the paper: U. Meyer, P. Sanders,
+ * $\Delta$-stepping: a parallelizable shortest path algorithm, Journal of Algorithms,
+ * Volume 49, Issue 1, 2003, Pages 114-152, ISSN 0196-6774,
+ * https://doi.org/10.1016/S0196-6774(03)00076-2.
+ *
+ * <p>
+ * The algorithm solves the single source shortest path problem in a graph with no
+ * negative weight edges. Its advantage of the {@link DijkstraShortestPath}
+ * algorithm is that it can benefit from multiple threads. While the Dijkstra`s
+ * algorithm is fully sequential and the {@link BellmanFordShortestPath} algorithm
+ * has high parallelism since all edges can be relaxed in parallel, the delta-stepping
+ * introduces parameter delta, which, when chooses optimally, yields still good parallelism
+ * and at the same time enables avoiding too many re-relaxations of the edges.
```

Comment:
Also I have not metioned any parallilisation details here yet, because I  will make changes to it in order to improve performance.

Reply:
ok sounds good. Forgot to mention: please include runtime and space complexity information here using big-O notation.